### PR TITLE
Structure placeholders in palette

### LINKED ISSUE #1647: Special map palette entries for placing structures
Currently, the only way to place a structure on a map is via the `placements` attribute in the `world` object.  However, this leads to blind reckoning when trying to "connect" a structure with other elements of the map; one has to manually count cells to determine the coordinate offset to supply in the `placements` attribute.  For example, in #1628 the road [here](https://github.com/swarm-game/swarm/pull/1628/files#r1400060997) is supposed to connect to a `rubbish enclosure`.

More convenient would be a way to insert a "reference cell" in the map grid that designates some corner of the structure to be placed.  This would be treated as a supplemental palette entry, like `mask` is currently.  This structure palette entry would specify which corner of the structure it represents: bottom-left, bottom-right, top-left, or top-right.

# History

Originally @byorgey suggested in https://github.com/swarm-game/swarm/issues/1138#issuecomment-1458793882:
> Sometimes you might just want to *draw* the placement of sub-maps on a map --- *i.e.* perhaps a single character in the palette could stand in for an entire map.

My thought at the time was https://github.com/swarm-game/swarm/issues/1138#issuecomment-1459075736:

> I thought about this, but I'm not convinced it's worth supporting.  Currently, map cells have 1:1 correspondence with entity or robot placement, but this would no longer hold true for multi-cell structures.  So the "visualization" benefit evaporates.  And trying to draw the complete overlay (edit: "footprint") of a submap onto a parent map leads to overspecification/maintainability problems.

However, I've now come around to the idea.

### PR DESCRIPTION
**Pull request by kostmo**
Closes #1647

```
scripts/play.sh -i data/scenarios/Testing/2086-structure-palette.yaml --autoplay
```

Screenshot from 2024-07-31 10-49-57

### CODE REVIEW

**Review by xsebek — APPROVED** [review]
@kostmo, this looks great! 👍 Was it only possible to use offsets until now? 

(As a disclaimer I have not used structures yet.)

File: src/swarm-scenario/Swarm/Game/Scenario/Topography/Cell.hs [hunk 1]
```diff
@@ -97,7 +97,7 @@ instance FromJSONE (TerrainEntityMaps, RobotMap) Cell where
         case meName of
           Nothing -> return ENothing
           Just "erase" -> return EErase
-          Just name -> fmap EJust . localE (view entityMap . fst) $ getEntity name
+          Just n -> fmap EJust . localE (view entityMap . fst) $ getEntity n
```
[the remark this answers is not in the document]
Comment (kostmo) [reply src/swarm-scenario/Swarm/Game/Scenario/Topography/Cell.hs:100]: This is to fix a name-shadowing warning due to the new `name` field.

**Comment by kostmo** [comment]
> Was it only possible to use offsets until now?

Before, it was still possible to place structures anywhere that we can now, by supplying explicit offset coordinates.  But it's easier to visualize now with palette markers.

File: src/swarm-scenario/Swarm/Game/Scenario/Topography/Cell.hs (hunk 1 above)
[the remark this answers is not in the document]
Comment (xsebek) [reply src/swarm-scenario/Swarm/Game/Scenario/Topography/Cell.hs:100]: @kostmo that is a very generic name `name` - could you import it with hiding the the name or qualified instead?

**Pull request merged**